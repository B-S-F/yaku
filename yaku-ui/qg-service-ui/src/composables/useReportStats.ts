// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

import type { MaybeRef } from '@vueuse/core'
import type { RunResultStatus } from '~/types'
import type { RunReportV1 } from '~/helpers'
import { unref } from 'vue'
import { getLargestRemainder } from '~/utils/getLargestRemainder'

type UseReportStatsReturn = {
  results: {
    GREEN: number
    YELLOW: number
    RED: number
    NA: number
    UNANSWERED: number
  }
  answered: {
    automatic: number
    manual: number
  }
}

export type UseReportStatsParams = {
  report: MaybeRef<RunReportV1>
}

export function* iterateOnChecks(chapters: RunReportV1['chapters']) {
  for (const { requirements } of chapters) {
    for (const r of requirements) {
      for (const c of r.checks) {
        // we do not validate down to the result to stay compatible with RunResultsV0 stats for now
        yield c
      }
    }
  }
}

const getValue = (m: Map<RunResultStatus, number>, k: RunResultStatus) =>
  m.get(k) ?? 0

const getAnsweredPourcentage = (
  report: RunReportV1,
): UseReportStatsReturn['answered'] => {
  let total = 0
  let automaticCount = 0,
    manualCount = 0

  for (const item of iterateOnChecks(report.chapters)) {
    total += 1
    if (item.type === 'Manual') {
      manualCount += 1
    } else if (item.type === 'Automation') {
      automaticCount += 1
    }
  }

  const raw = {
    automatic: (automaticCount / total) * 100,
    manual: (manualCount / total) * 100,
  }
  const [automatic, manual] = getLargestRemainder(
    [raw.automatic, raw.manual],
    raw.automatic + raw.manual,
  )

  return {
    automatic,
    manual,
  }
}

const getResultsPourcentage = (
  report: RunReportV1,
): UseReportStatsReturn['results'] => {
  let total = 0
  const acc = new Map<RunResultStatus, number>()

  for (const item of iterateOnChecks(report.chapters)) {
    acc.set(
      item.status as RunResultStatus,
      (acc.get(item.status as RunResultStatus) ?? 0) + 1,
    )
    total += 1
  }

  const greenAmount = getValue(acc, 'GREEN')
  const yellowAmount = getValue(acc, 'YELLOW')
  const redAmount =
    getValue(acc, 'RED') + getValue(acc, 'FAILED') + getValue(acc, 'ERROR')
  const naAmount = getValue(acc, 'NA')
  const unansweredAmount =
    getValue(acc, 'UNANSWERED') + getValue(acc, 'SKIPPED')

  const rawGreen = (greenAmount / total) * 100
  const rawYellow = (yellowAmount / total) * 100
  const rawRed = (redAmount / total) * 100
  const rawNa = (naAmount / total) * 100
  const rawUnanswered = (unansweredAmount / total) * 100

  const [GREEN, YELLOW, RED, NA, UNANSWERED] = getLargestRemainder(
    [rawGreen, rawYellow, rawRed, rawNa, rawUnanswered],
    100,
  )

  return {
    GREEN,
    YELLOW,
    RED,
    NA,
    UNANSWERED,
  }
}

export const useReportStats = ({
  report,
}: UseReportStatsParams): UseReportStatsReturn => {
  // 2 passes on the object seems fine for now
  const answered = getAnsweredPourcentage(unref(report))
  const results = getResultsPourcentage(unref(report))

  return {
    results,
    answered,
  }
}
