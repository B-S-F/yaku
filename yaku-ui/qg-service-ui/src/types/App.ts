// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

export type AppParameter = {
  name: string
  description?: string
  example?: string
  optional?: boolean
}

export type App = {
  name: string
  envs: AppParameter[]
}
