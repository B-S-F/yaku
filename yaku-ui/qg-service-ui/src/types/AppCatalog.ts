// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

export type App = {
  id: string
  name: string
  executableName: string
  description?: string
  handbookLink: string
  version: string
  parameters: Parameter[]
  configFile?: ConfigFile
  creationTime: string
  lastModificationTime: string
}

export type BaseParameter<T extends string> = {
  name: string
  type: 'env' | 'arg'
  optional: boolean
  isSecret: boolean
  content: T
}

export type EnumParameter = BaseParameter<'enum'> & { contentEnum: string[] }
export type StringParameter = BaseParameter<'string'>
export type BooleanParameter = BaseParameter<'boolean'>

export type Parameter = EnumParameter | StringParameter | BooleanParameter

export type ConfigFile = {
  name: string
  example: string
  schema: string
  format: 'yaml'
}
