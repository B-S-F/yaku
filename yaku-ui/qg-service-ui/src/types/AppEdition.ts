// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

import type { App, Parameter } from './AppCatalog'

export type ParameterValue = {
  parameter: Parameter
  value: string
}

export type AppFilled = {
  app: App
  args: ParameterValue[]
  envs: ParameterValue[]
}
