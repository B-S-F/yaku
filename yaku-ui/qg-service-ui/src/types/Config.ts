// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

export type Config = {
  id: number
  name: string
  description?: string
  creationTime: string
  lastModificationTime: string
  files: {
    qgConfig?: string
    qgAnswersSchema?: string
    additionalConfigs?: string[]
  }
}
