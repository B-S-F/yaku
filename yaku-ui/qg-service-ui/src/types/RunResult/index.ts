// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

export * from './RunResultV0'
export * from './RunResultV1'
export type { RunResultStatus, ManualStatus } from './Common'

import type { RunResultV0 } from './RunResultV0'
import type { RunResultV1 } from './RunResultV1'
export type RunResult = RunResultV0 | RunResultV1
