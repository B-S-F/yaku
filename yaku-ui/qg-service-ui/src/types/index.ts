// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

export * from './App'
export * from './Autopilot'
export * from './Config'
export * from './Dialog'
export * from './Editor'
export * from './Environment'
export * from './Error'
export * from './Finding'
export * from './Items'
export * from './Onboarding'
export * from './OssComplianceInformation'
export * from './Run'
export * from './RunReport'
export * from './Secret'
export * from './StatusPillDisplay'
