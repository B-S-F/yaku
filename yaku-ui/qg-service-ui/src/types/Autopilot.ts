// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

import { App } from './App'

export type Autopilot = {
  name: string
  description: string
  apps: App[]
}

export type AutopilotSnippet = {
  run: string
  env: Record<string, string>
}
