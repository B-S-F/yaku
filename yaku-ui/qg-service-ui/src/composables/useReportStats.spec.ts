// SPDX-FileCopyrightText: 2024 grow platform GmbH
//
// SPDX-License-Identifier: MIT

import { describe } from 'vitest'
import { useReportStats } from './useReportStats'
import { load } from 'js-yaml'
import runResultRed from '~/mocks/fixtures/qgResultV0/qg-result--red.yaml?raw'
import runResultGreen from '~/mocks/fixtures/qgResultV0/qg-result--green.yaml?raw'
import runResultUnanswered from '~/mocks/fixtures/qgResultV0/qg-result--unanswered.yaml?raw'
import { RunResultV0 } from '~/types/RunResult'
import { convertRunResultToReport } from '~helpers'

describe('useReportStats', () => {
  it('returns the expected values with a successful run', () => {
    const runResult = load(runResultGreen) as RunResultV0
    const report = convertRunResultToReport(runResult)!.report
    const stats = useReportStats({ report })
    expect(stats).toMatchSnapshot()
  })

  it('returns the expected values with a failed run', () => {
    const runResult = load(runResultRed) as RunResultV0
    const report = convertRunResultToReport(runResult)!.report
    const stats = useReportStats({ report })
    expect(stats).toMatchSnapshot()
  })

  it('returns the expected values with 100% unanswered', () => {
    const runResult = load(runResultUnanswered) as RunResultV0
    const report = convertRunResultToReport(runResult)!.report
    const stats = useReportStats({ report })
    expect(stats).toMatchSnapshot()
  })
})
