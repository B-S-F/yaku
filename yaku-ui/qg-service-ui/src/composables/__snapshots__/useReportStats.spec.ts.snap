// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useReportStats > returns the expected values with 100% unanswered 1`] = `
{
  "answered": {
    "automatic": 100,
    "manual": 0,
  },
  "results": {
    "GREEN": 0,
    "NA": 0,
    "RED": 0,
    "UNANSWERED": 100,
    "YELLOW": 0,
  },
}
`;

exports[`useReportStats > returns the expected values with a failed run 1`] = `
{
  "answered": {
    "automatic": 89,
    "manual": 11,
  },
  "results": {
    "GREEN": 46,
    "NA": 4,
    "RED": 28,
    "UNANSWERED": 18,
    "YELLOW": 4,
  },
}
`;

exports[`useReportStats > returns the expected values with a successful run 1`] = `
{
  "answered": {
    "automatic": 86,
    "manual": 14,
  },
  "results": {
    "GREEN": 89,
    "NA": 11,
    "RED": 0,
    "UNANSWERED": 0,
    "YELLOW": 0,
  },
}
`;
